lib/gssapi/mech/gss_duplicate_cred.c

```diff
@@ -31,52 +31,84 @@
 #include "mech_locl.h"
 
 static OM_uint32
-_gss_copy_cred_element(OM_uint32 *minor_status,
-                       struct _gss_mechanism_cred *mc,
-                       struct _gss_mechanism_cred **out)
+copy_cred_element(OM_uint32 *minor_status,
+                  struct _gss_mechanism_cred *mc,
+                  struct _gss_mechanism_cred **out)
 {
     gssapi_mech_interface m = mc->gmc_mech;
-    OM_uint32 major_status;
-    gss_name_t name;
-    gss_cred_id_t cred;
+    OM_uint32 major_status, tmp;
+    struct _gss_mechanism_name mn;
+    struct _gss_mechanism_cred *new_mc;
     OM_uint32 initiator_lifetime, acceptor_lifetime;
     gss_cred_usage_t cred_usage;
+    gss_cred_id_t dup_cred = GSS_C_NO_CREDENTIAL;
 
-    if (m->gm_duplicate_cred)
-        return m->gm_duplicate_cred(minor_status, (gss_const_cred_id_t)mc,
-                                    (gss_cred_id_t *)out);
+    if (m->gm_duplicate_cred) {
+        major_status = m->gm_duplicate_cred(minor_status,
+					    mc->gmc_cred, &dup_cred);
+    } else if (m->gm_import_cred && m->gm_export_cred) {
+	gss_buffer_desc export;
 
-    /* This path won't work for ephemeral creds */
+	major_status = m->gm_export_cred(minor_status, mc->gmc_cred, &export);
+	if (major_status == GSS_S_COMPLETE) {
+	    major_status = m->gm_import_cred(minor_status, &export, &dup_cred);
+	    gss_release_buffer(&tmp, &export);
+	}
+    } else
+	major_status = GSS_S_UNAVAILABLE;
+
+    if (major_status != GSS_S_UNAVAILABLE) {
+	if (dup_cred != GSS_C_NO_CREDENTIAL) {
+	    new_mc = calloc(1, sizeof(*new_mc));
+	    if (new_mc == NULL) {
+		*minor_status = ENOMEM;
+		m->gm_release_cred(&tmp, &dup_cred);
+		return GSS_S_FAILURE;
+	    }
+
+	    new_mc->gmc_mech = m;
+	    new_mc->gmc_mech_oid = mc->gmc_mech_oid;
+	    new_mc->gmc_cred = dup_cred;
+
+	    *out = new_mc;
+	}
+
+	return major_status;
+    }
+
+    mn.gmn_mech = m;
+    mn.gmn_mech_oid = mc->gmc_mech_oid;
+    mn.gmn_name = GSS_C_NO_NAME;
+
+    /* This path won't work for ephemeral creds or cred stores */
```

Comment:
This comment was in the original.  Why doesn't this work for ephemeral creds or cred stores?  I think it does.

Reply:
Ah, right, it is correct in the export/import case, but not in the duplicate case.  Just move the comment back up.